agent-control/src/secret_retriever/on_host/retrieve.rs

```diff
@@ -0,0 +1,196 @@
+use crate::agent_control::config::OpAMPClientConfig;
+use crate::agent_control::defaults::AUTH_PRIVATE_KEY_FILE_NAME;
+use crate::agent_control::run::BasePaths;
+use crate::opamp::auth::config::ProviderConfig;
+use crate::secret_retriever::OpampSecretRetriever;
+use crate::secrets_provider::SecretsProvider;
+
+/// Helper struct to determine the path and retrieve the secret using the File provider.
+pub struct OnHostSecretRetriever<P> {
+    pub base_paths: BasePaths,
+    pub provider: P,
+}
+#[derive(Debug, thiserror::Error)]
+#[error("{0}")]
+pub struct OnHostRetrieverError(String);
+impl<P> OnHostSecretRetriever<P>
+where
+    P: SecretsProvider,
+{
+    pub fn new(base_paths: BasePaths, provider: P) -> Self {
+        Self {
+            base_paths,
+            provider,
+        }
+    }
+}
+
+impl<P> OpampSecretRetriever for OnHostSecretRetriever<P>
+where
+    P: SecretsProvider,
+{
+    type Error = OnHostRetrieverError;
+
+    fn retrieve(&self, opamp_config: &OpAMPClientConfig) -> Result<String, Self::Error> {
+        let path_buf = if let Some(auth_config) = &opamp_config.auth_config {
+            if let Some(ProviderConfig::Local(local_config)) = &auth_config.provider {
+                local_config.private_key_path.clone()
+            } else {
+                self.base_paths.local_dir.join(AUTH_PRIVATE_KEY_FILE_NAME)
+            }
+        } else {
+            self.base_paths.local_dir.join(AUTH_PRIVATE_KEY_FILE_NAME)
+        };
+
+        let secret_path = path_buf.to_string_lossy().to_string();
+
+        self.provider
+            .get_secret(&secret_path)
+            .map_err(|e| OnHostRetrieverError(format!("Failed to retrieve file secret: {}", e)))
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::opamp::auth::config::{AuthConfig, LocalConfig};
+    use crate::opamp::client_builder::PollInterval;
+    use http::HeaderMap;
+    use std::fmt;
+    use std::path::PathBuf;
+    use url::Url;
+
+    struct MockProvider {
+        expected_path_tocheck: String,
+        should_succeed: bool,
+    }
+
+    #[derive(Debug)]
+    struct MockError;
+    impl fmt::Display for MockError {
+        fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+            write!(f, "MockError")
+        }
+    }
+    impl std::error::Error for MockError {}
+
+    impl SecretsProvider for MockProvider {
```

Comment:
Same here, why not [mockall::mock!](https://docs.rs/mockall/latest/mockall/macro.mock.html)?